v1.16 Backports 2024-08-27

tklauser opened the pull request:
* [x] #34398 (@mhofstetter)
 * [ ] #34295 (@MrFreezeex)
 * [x] #34221 (@jschwinger233)
 * [x] #34476 (@gandro)
 * [x] #34453 (@pchaigno)
   * :warning: conflict due to `LB6_SKIP_MAP` not being present on `v1.16`. Resolved by dropping the respective hunk of the patch (i.e. https://github.com/cilium/cilium/pull/34453/commits/71b9cdef49bc42688535f98ec59e9dc09ce5c6b8).
 * [x] #34298 (@julianwiedmann)
 * [x] #34389 (@christarazi)
 * [ ] #33805 (@jschwinger233)
 * [x] #34474 (@sayboras)
 * [x] #34249 (@joamaki)
 * [x] #33896 (@aanm)
 * [ ] #34279 (@alex-berger)
 * [x] #34276 (@harsimran-pabla)

Once this PR is merged, a GitHub action will update the labels of these PRs:
```upstream-prs
 34398 34295 34221 34476 34453 34298 34389 33805 34474 34249 33896 34279 34276
```

aanm committed:
pkg/endpointmanager: don't hold lock while iterating over subscribers

[ upstream commit d84274428103b854ab0b3ec7340e2d8626c5ebca ]

Holding the lock while iterating over the subscribers is causing a
potential deadlock with pkg/auth. Since pkg/auth is a subscriber of the
endpoint manager, it will lock an internal mutex and then perform a
GetEndpoints which will also perform a RLock to the endpointsmanager's
Mutex. This itself is not an issue. However, on a separate go routine
the pkg/auth will perform a lock on its internal mutex and then will do
a GetEndpoints, causing inconsistent locking between these two cases
which can cause a deadlock in the agent.

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

jschwinger233 committed:
ipsec: Document a new cause of XfrmInStateProtoError

[ upstream commit 2387d597752132e1f2fc351f11bc04c2d76ea6e9 ]

Signed-off-by: gray <gray.liang@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

joamaki committed:
daemon: Fix parsing of complex api-rate-limit arguments

[ upstream commit 82962533c31fd4d0a7839cfd09e4cb08658c658e ]

In d7ac88839768 the API rate limit parsing was redone to use StringToString.
This broke parsing of api-rate-limits of form:

  "foo=rate-limit:1/s,rate-burst:1,bar=rate-limit:2/s"

The pflag StringToString failed to split this correctly leading to
it trying to parse "rate-burst:1" as a "key=value". Additionally
the change broke the support for specifying the rate limits in JSON
format.

Expand the test-cases to cover this usage and fix the issue by switching
back to parsing with the command.ToStringMapStringE method.

Fixes: #31894
Fixes: d7ac88839768 ("daemon: Move rate limiter to daemon/restapi")

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

harsimran-pabla committed:
bgpv2: correct error message log

[ upstream commit 1728dc479026984e72c73a0e7f8ac9d2114d530a ]

Missing service option can be user error, it is not a bug - so removing
BUG prefix to avoid unnecessary confusion.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

MrFreezeex committed:
helm: also mount kvstoremesh-specific certificate into cilium-operator

[ upstream commit 3f5c6993f1949ab8532edff656108dabea2ec159 ]

In commit c464e66620a45f56f207d923f7d8f843e59b813e the
kvstoremesh-specific certificate was mounted into cilium-agent. Since
the recent addition of endpointslicesync the operator needs to connect to
clustermesh components too meaning that the clustermesh config and
secrets should be mirrored to the operator deployment as well.
With the addition of the kvstoremesh secrets to the agent and not the operator
this resulted into the code connecting to the kvstoremesh not working in the
operator context which is now the default as well.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

julianwiedmann committed:
sysctl: allow uppercase letters

[ upstream commit 0f13b32d0445e084249762cf6b315bd9f8d94529 ]

sysctl names potentially contain uppercase letters (a typical example would
be network interface names). Support them in the regex.

Fixes: https://github.com/cilium/cilium/issues/34287
Signed-off-by: Julian Wiedmann <jwi@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

jschwinger233 committed:
bpf: Add revdnat_only parameter to handle_nat_fwd()

[ upstream commit 7a97965db9292ef6f3b944cb3d17dd6cbcdfa954 ]

This commit doesn't change code logic, merely adds a new parameter to
handle_nat_fwd().

The new parameter `revdnat_only` can be used to skip fib_lookup_redirect
and bpf_snat in handle_nat_fwd(). This is useful when IPsec/wireguard is
enabled with bpf nodeport: we only want to perform revDNAT for reply
traffic before encryption.

Signed-off-by: gray <gray.liang@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

jschwinger233 committed:
bpf: revDNAT reply traffic before IPsec encryption

[ upstream commit f6eb2f4a4771d9b5411b8e458bbe2921714e063f ]

When IPsec is enabled with L7 ingress policy, nodeport, and KPR, the
reply path for a pod-to-remote-hostport connection is supposed to be:
from-container@lxc -> L7 proxy -> from-host@cilium_host ->
to-host@cilium_net -> xfrm IPsec encryption -> to-netdev@eth0.

The problem of above path is, we don't stand a chance to perform revDNAT.
It is supposed to take place at to-netdev@eth0, but bpf code fails to do
so because datapath sees IPsec-encrypted ESP.

Another existing revDNAT position is from-container@lxc, but we intend to
remove that in future and rely on node's egress interface only
(to-netdev@eth0). On the other hand, revDNAT at from-container@lxc still
doesn't work in this L7 proxy case because proxy_redirect happens prior
to the revDNAT.

This commit adds a new revDNAT hook at to-host@cilium_net for IPsec
specifically, where handle_nat_fwd() is called with revdnat_only=1 to
skip unnecessary fib_lookup, redirect, and snat.

Fixes: https://github.com/cilium/cilium/issues/32897

Signed-off-by: gray <gray.liang@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

mhofstetter committed:
cilium-dbg: add command 'envoy admin config'

[ upstream commit 27cd07eb715826d180f26a5abde7b7abf2653a72 ]

Currently, dumping the applied envoy config requires either exposing
the envoy admin interface via IP listener and accessing it via browser
or SSHing into the cilium-agent, installing `curl` and accessing the
envoy admin interface via UNIX socket (`/var/run/cilium/envoy/sockets/admin.sock`).

Since both approaches aren't really userfriendly (e.g. during development), this
commit introduces a new command `envoy admin config` to the `cilium-dbg` binary.
This way, the Envoy Config of a Cilium Node can be viewed without having to install
curl and remember the UNIX socket path.

In addition to viewing the config it's also possible to

* filter by resource type by appending it to the command
  (e.g. `cilium-dbg envoy admin config listeners`)
* filter by resource name
  (e.g. `cilium-dbg envoy admin config -n test`)
  (Can be combined with resource type filtering)

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

mhofstetter committed:
cilium-dbg: add command 'envoy admin serverinfo'

[ upstream commit 69824b6f6596d9629b952f2108f88a3fcd8eecbd ]

Currently, viewing the serverinfo of Envoy via the admin interface requires
either exposing the envoy admin interface via IP listener and accessing it via browser
or SSHing into the cilium-agent, installing `curl` and accessing the
envoy admin interface via UNIX socket (`/var/run/cilium/envoy/sockets/admin.sock`).

Since both approaches aren't really userfriendly (e.g. during development), this commit
introduces a new command `envoy admin serverinfo` to the `cilium-dbg` binary. This way,
the Envoy Config of a Cilium Node can be viewed without having to install curl and
remember the UNIX socket path.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

mhofstetter committed:
cilium-dbg: add command 'envoy admin metrics'

[ upstream commit 600c26e6ce6fa9a0c0aed9cfcab1b976df9b1ef2 ]

Currently, viewing the Prometheus metrics of Envoy via the admin interface requires
either setting up prometheus integration, exposing the envoy admin interface via
IP listener and accessing it via browser
or SSHing into the cilium-agent, installing `curl` and accessing the
envoy admin interface via UNIX socket (`/var/run/cilium/envoy/sockets/admin.sock`).

Since both approaches aren't really userfriendly (e.g. during development), this commit
introduces a new command `envoy admin metrics` to the `cilium-dbg` binary. This way,
the Envoy Config of a Cilium Node can be viewed without having to install curl and
remember the UNIX socket path.

In addition it's also possible to filter the metrics by using the `-f` flag.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

mhofstetter committed:
cilium-dbg: add command 'envoy admin clusters'

[ upstream commit 2386a097eafa02b6306583b4a5bed5b803102082 ]

Currently, viewing the clusters of Envoy via the admin interface requires
either exposing the envoy admin interface via IP listener and accessing it via browser
or SSHing into the cilium-agent, installing `curl` and accessing the
envoy admin interface via UNIX socket (`/var/run/cilium/envoy/sockets/admin.sock`).

Since both approaches aren't really userfriendly (e.g. during development), this commit
introduces a new command `envoy admin clusters` to the `cilium-dbg` binary. This way,
the Envoy Config of a Cilium Node can be viewed without having to install curl and
remember the UNIX socket path.

Note: By default, the output is in "structured textformat", it's also possible to get
the JSON output via `-o json`.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

mhofstetter committed:
cilium-dbg: add command 'envoy admin listeners'

[ upstream commit 8dd1f6d029636534779aea49afddc0cb32c911e1 ]

Currently, viewing the listeners of Envoy via the admin interface requires
either exposing the envoy admin interface via IP listener and accessing it via browser
or SSHing into the cilium-agent, installing `curl` and accessing the
envoy admin interface via UNIX socket (`/var/run/cilium/envoy/sockets/admin.sock`).

Since both approaches aren't really userfriendly (e.g. during development) , this commit
introduces a new command `envoy admin listeners` to the `cilium-dbg` binary. This way,
the Envoy Config of a Cilium Node can be viewed without having to install curl and
remember the UNIX socket path.

Note: By default, the output is in "structured textformat", it's also possible to get
the JSON output via `-o json`.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

mhofstetter committed:
cilium-dbg: add command 'envoy admin certs'

[ upstream commit 98635cedbfea52305c04467d3030d0b01b5c6c8b ]

Currently, viewing the certificates of Envoy via the admin interface requires
either exposing the envoy admin interface via IP listener and accessing it via browser
or SSHing into the cilium-agent, installing `curl` and accessing the
envoy admin interface via UNIX socket (`/var/run/cilium/envoy/sockets/admin.sock`).

Since both approaches aren't really userfriendly (e.g. during development), this commit
introduces a new command `envoy admin certs` to the `cilium-dbg` binary. This way,
the Envoy Config of a Cilium Node can be viewed without having to install curl and
remember the UNIX socket path.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

christarazi committed:
daemon: Fix error logic flow for pod store being out of date

[ upstream commit 143ca65aeeaeb6b82aa951ef1c1ee390da4290a8 ]

In the endpoint creation path, if Cilium does not have the K8s Pod
reference in the local store, then the logic is to fetch the latest from
the apiserver directly. However, in the case that the fetch succeeds,
the error variable is not clear. This results in Cilium continuing in
the "unhappy" path, even though it should be the "happy" path.

Relevant log msgs from a sysdump displaying this behavior:

```
level=info msg="Create endpoint request" addressing="&{10.0.0.133 ba039cfd-7061-4c45-8526-1a5f569d16de default   }" containerID=3bed19eb3aabbb100ec39037a364cf1cbb10ca2666c14faec9893a2dc129844a containerInterface=eth0 datapathConfiguration="&{false false false false false <nil>}" interface=lxc5c692ead6e7a k8sPodName=default/nginx-static-pod-master-node k8sUID=cc8a369e7eac1fc96b6e3b51830c86e9 labels="[]" subsys=daemon sync-build=true
level=warning msg="Detected outdated Pod UID during Endpoint creation. Endpoint creation cannot proceed with an outdated Pod store. Attempting to fetch latest Pod." k8sPodName=default/nginx-static-pod-master-node k8sUID=cc8a369e7eac1fc96b6e3b51830c86e9 subsys=daemon
level=warning msg="Timeout occurred waiting for Pod store, fetching latest Pod via the apiserver." k8sPodName=default/nginx-static-pod-master-node k8sUID=cc8a369e7eac1fc96b6e3b51830c86e9 subsys=daemon
level=warning msg="Unable to fetch kubernetes labels" ciliumEndpointName=/ containerID= containerInterface= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=0 error="pod store outdated" ipv4= ipv6= k8sPodName=/ subsys=api
```

Fixes: f6606c6ccd ("daemon,endpoint,cni: Pass pod UID through CNI ADD")
Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

pchaigno committed:
lbmap: Create SKIP_LB maps unconditionally

[ upstream commit 2321e14730815dd9bcc1f82c2095ebf0d670955a ]

or at least not conditioned on LRP being enabled. On the BPF side, these
maps are created regardless of whether LRP is enabled. We should do the
same on the agent side, to ensure that the agent always creates those
maps early on startup, before they are created as part of program
loading. Creating maps as part of program loading can lead to errors due
to a race condition.

Fixes: 6496821001 ("daemon: Create IPsec and LRP maps early on startup")
Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

pchaigno committed:
datapath: Set LB6_SKIP_MAP

[ upstream commit 547f1e312a8976144979a6914352ddded41b38a8 ]

The LB{4,6}_SKIP_MAP were defined in commit 33552812c3f ("pkg/maps: Add
new maps to skip service load-balancing"), but only the IPv4 macro was
defined by the agent. As a result, a map named literally LB6_SKIP_MAP
was created at program load time, whereas the agent expected the map to
be named cilium_skip_lb6.

Thankfully, this didn't lead to any issues other than a few error
messages at load time due to a race condition. This race condition
arises when several BPF programs try to create the same map
concurrently. It is avoided by having the agent create the maps ahead of
loading. In this case, it didn't work because the agent would create the
map under a different name.

This commit fixes it.

Fixes: 33552812c3f ("pkg/maps: Add new maps to skip service load-balancing")
Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

sayboras committed:
ingress: Avoid opening of port 80 for TLSPassthrough only

[ upstream commit 2f09c729081182fca7307a7df3d9758a916f8561 ]

In TLSPassthrough only, port 80 is not required to be exposed in LB
service.

Signed-off-by: Tam Mach <tam.mach@cilium.io>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

gandro committed:
bpf: Fix `Prune` map operation leaking BPF map entries

[ upstream commit cef0b371c841b457f98e4b184af74589dcffb23a ]

This fixes a bug in the `mapOps.Prune` implementation where it leaked
entries during the prune operation. The cause for this was that the
`Prune` implementation deleted entries while iterating over them,
causing it to skip iterations at least in LPM tries.

This commit fixes the issue by first collecting all keys to delete, and
then deleting them in a second phase. A unit test which produces the
faulty behavior before the fix is also added to ensure we do not regress
on this.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

alex-berger committed:
fix: add regression tests for api-rate-limit validation

[ upstream commit e875f088e35b6279fb7f538c88ba33c108a35089 ]

Signed-off-by: Alexander Berger <alex-berger@gmx.ch>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

alex-berger committed:
fix: support validation of stringToString values in ConfigMap

[ upstream commit c2f1988098afc84ef218b8956721c19906bd4b47 ]

Fixes:#33095

This is the second attempt to fix https://github.com/cilium/cilium/issues/33095,
the first attempt lead to regression https://github.com/cilium/cilium/issues/34129
and was reverted in https://github.com/cilium/cilium/pull/34277.

Using command.ToStringMapStringE(...) instead of cast.ToStringMapStringE(...)

Signed-off-by: Alexander Berger <alex-berger@gmx.ch>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

sayboras reviewed (approved):
Thanks and looks good for my commit ✅

MrFreezeex reviewed (approved):
Thanks! My commit looks good

mhofstetter reviewed (approved):
My commits look good! Thanks!

pchaigno commented on bpf/node_config.h:
```diff
@@ -176,6 +176,7 @@ DEFINE_IPV6(HOST_IP, 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0xa, 0x
 #define LB4_MAGLEV_MAP_OUTER test_cilium_lb4_maglev_outer
 #define LB6_MAGLEV_MAP_OUTER test_cilium_lb6_maglev_outer
 #define LB4_SKIP_MAP test_cilium_skip_lb4
+#define LB6_SKIP_MAP test_cilium_skip_lb6
```
> * ⚠️ conflict due to `LB6_SKIP_MAP` not being present on `v1.16`. Resolved by dropping the respective hunk of the patch (i.e. [71b9cde](https://github.com/cilium/cilium/commit/71b9cdef49bc42688535f98ec59e9dc09ce5c6b8)).

It's actually there, but it's hidden in `bpf_sock.c` instead of `lib/lb.h`.

pchaigno reviewed (changes requested):
Third commit can be backported but the target code moved.

pchaigno committed:
bpf: Do not prealloc LB6_SKIP_MAP map

[ upstream commit 992e36be93ecad7af29b0659049d828e8bcd2ec9 ]

[ backporter's note: LB6_SKIP_MAP got moved from bpf/bpf_sock.c to
  bpf/lib/lb.h between v1.16 and main. Adjusted patch accordingly. ]

The previous commit fixed the agent to create the LB6_SKIP_MAP map on
startup. That uncovered a mismatch between the agent and datapath
definitions of the map. Contrary to its IPv4 variant, it seems the
NO_PREALLOC flag was forgotten on the datapath definition of
LB6_SKIP_MAP.

This commit fixes it.

Fixes: ad09e20df402 ("bpf: Skip LB for local redirect svc tuples")
Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

pchaigno reviewed (approved):
Thanks!
